uamqp/async_ops/client_async.py

```diff
@@ -840,8 +840,13 @@ async def _client_run_async(self):
             if self._timeout > 0:
                 timespan = now - self._last_activity_timestamp
                 if timespan >= self._timeout:
-                    _logger.info("Timeout reached, closing receiver.")
-                    self._shutdown = True
+                    self._timeout_reached = True
+                    if self._shutdown_after_timeout:
```

Comment:
I don't see a docstring for this being passed in or defined anywhere?

Reply:
it's in client.py:
`self._shutdown_after_timeout = kwargs.pop('shutdown_after_timeout', True)`

TBH, I didn't plan to make this parameter public, because it seems to have a conflict with the current `timeout` param on the client:
`:param timeout: A timeout in milliseconds. The receiver will shut down if no
     new messages are received after the specified timeout. If set to 0, the receiver
     will never timeout and will continue to listen. The default is 0.
`

or we could make it public but add some additional notes in the `timeout`? like:
`:param timeout: A timeout in milliseconds. The receiver will shut down **if shutdown_after_timeout is set to true** and no
     new messages are received after the specified timeout. If set to 0, the receiver
     will never timeout and will continue to listen. The default is 0.
`

@annatisch do you think we shall make this parameter public?

Reply:
Seems like it would be okay to make this public.

Reply:
My only though is that I'm not sure how well this plays with the other receive mechanisms. Though it makes sense for the iterator. I'm wondering - rather than make this a top-level client parameter - maybe we could only expose it at the iterator function?

Reply:
@annatisch This is a pretty good point of view which pushes me to think harder on this.

The right thing to do here from my point after revisiting the code base is: **It should be a top-level client parameter. And to your concern, we should ensure the parameter plays well with other receive mechanisms -- write tests for them.**

I should admit  the idea of only exposing it at the iterator looks the most simplistic. However:
  - Our goal is to keep the uamqp receiver alive in ServiceBus world which I don't think iterator only would help us achieve.
    - uamqp has three receiving mechanisms in total: `pull/push/iterator`.
    - Service bus uses `pull` and `iterator`.
      - Keeping uamqp receiver alive (not shutdown after timeout) should work for both `pull` and `iterator` mode.
      - If it's iterator only, then it means pull receive would shutdown the uamqp receiver after timeout.
  - Consistency concern: The key code to impact the behavior of "shutdown after timeout"  is in `do_work()` which is the base of all receiving mechanisms.
    - Making it generator only would bring extra complexity than it looks like, in which extra variables or flow need to be introduced to guarantee the consistency of receiver status when multiple receiving mechanisms are used simultaneously.

So overall, I believe making it a top-level parameter brings us more benefits than the costs as this meets our goal in SB and keeps the receiving flow clean and consistent in all ways.

Reply:
@annatisch , I have cover all the receiving mechanisms to work with new parameters and tests are added.